pkg/networkservice/mechanisms/memif/server.go

```diff
@@ -18,31 +18,62 @@ package memif
 
 import (
 	"context"
+	"net/url"
 
 	"git.fd.io/govpp.git/api"
 	"github.com/golang/protobuf/ptypes/empty"
+
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	memifMech "github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/memif"
 
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/metadata"
+
+	"github.com/networkservicemesh/sdk-vpp/pkg/tools/ifindex"
 )
 
 type memifServer struct {
-	vppConn api.Connection
+	vppConn            api.Connection
+	directMemifEnabled bool
 }
 
 // NewServer provides a NetworkServiceServer chain elements that support the memif Mechanism
-func NewServer(vppConn api.Connection) networkservice.NetworkServiceServer {
-	return &memifServer{
-		vppConn: vppConn,
+func NewServer(vppConn api.Connection, options ...Option) networkservice.NetworkServiceServer {
+	m := &memifServer{
+		vppConn:            vppConn,
+		directMemifEnabled: true,
+	}
+
+	for _, opt := range options {
+		opt(m)
 	}
+
+	return m
 }
 
 func (m *memifServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
 	conn, err := next.Server(ctx).Request(ctx, request)
 	if err != nil {
 		return nil, err
 	}
+
+	if mechanism := memifMech.ToMechanism(conn.GetMechanism()); mechanism != nil {
+		// Direct memif if applicable
+		if memifSocketAddDel, ok := load(ctx, true); ok && m.directMemifEnabled {
+			_, ok := ifindex.Load(ctx, true)
+			if ok {
+				if err := del(ctx, conn, m.vppConn, true); err != nil {
+					_, _ = m.Close(ctx, conn)
+					return nil, err
+				}
+				mechanism.SetSocketFileURL((&url.URL{Scheme: memifMech.SocketFileScheme, Path: memifSocketAddDel.SocketFilename}).String())
+				delete(ctx, true)
+				ifindex.Delete(ctx, true)
+				return conn, nil
+			}
+		}
+	}
```

Comment:
I would be suggesting to not add complexity into memif server element and create a separate chain element for directmemif.

I'm fine with these changes if @edwarnicke has no comments on this.

Reply:
I'm also inclined to a new chain element for directmemif ... but if that is not possible, lets at least default the directMemif=false

Reply:
ok, changed default value